File: modules/student-experience/src/main/java/com/campusos/portal/service/PresentationService.java
package com.campusos.portal.service;

import com.campusos.portal.domain.*;
import com.campusos.portal.engine.TransitionMatrix;
import com.campusos.portal.engine.Transition;
import com.campusos.portal.engine.WorkflowSpec;
import com.campusos.portal.payload.Artifact;
import com.campusos.portal.payload.PayloadCodec;
import com.campusos.portal.payload.RequestPayload;
import com.campusos.portal.repo.RequestHistoryRepository;
import com.campusos.portal.repo.StudentRepository;
import com.campusos.portal.view.ActionButton;
import com.campusos.portal.view.DisplayLabels;
import com.campusos.portal.view.TrailEntry;
import com.campusos.portal.view.RequestCard;
import com.campusos.portal.view.TimelineStep;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.springframework.stereotype.Service;

/**
 * Turns a Request of ANY type into one RequestCard, using only WorkflowSpec metadata and
 * the payload's own polymorphic title()/subtitle()/artifacts(). No switch on RequestType.
 */
@Service
public class PresentationService {

    private static final DateTimeFormatter FMT =
            DateTimeFormatter.ofPattern("d MMM, HH:mm").withZone(ZoneId.systemDefault());

    private final PayloadCodec codec;
    private final RequestHistoryRepository history;
    private final StudentRepository students;

    public PresentationService(PayloadCodec codec, RequestHistoryRepository history,
                               StudentRepository students) {
        this.codec = codec;
        this.history = history;
        this.students = students;
    }

    public List<RequestCard> cards(Scope scope, List<Request> requests) {
        String studentName = studentName(scope);
        List<RequestCard> out = new ArrayList<>();
        for (Request r : requests) out.add(card(scope, r, studentName));
        return out;
    }

    public RequestCard card(Scope scope, Request r) {
        return card(scope, r, studentName(scope));
    }

    /** The student reads their own name in the trail, not the word "STUDENT". */
    private String studentName(Scope scope) {
        return students.findByIdAndTenantId(scope.studentId(), scope.tenantId())
                .map(s -> s.name).orElse("You");
    }

    private RequestCard card(Scope scope, Request r, String studentName) {
        WorkflowSpec spec = TransitionMatrix.spec(r.type);
        RequestPayload payload = codec.read(r.type, r.payload);
        List<RequestHistory> rows = history.findByRequestIdAndTenantIdAndStudentIdOrderByIdAsc(
                r.id, scope.tenantId(), scope.studentId());

        return new RequestCard(
                r.id,
                r.type.name(),
                spec.label(),
                payload.title(),
                payload.subtitle(),
                r.state.name(),
                DisplayLabels.stateLabel(spec.labelFor(r.state), payload.handledBy()),
                tone(spec, r.state),
                headline(spec, r, payload, rows),
                steps(spec, r, rows),
                studentAction(spec, r),
                artifacts(payload.artifacts()),
                FMT.format(r.createdAt),
                FMT.format(r.updatedAt),
                trail(rows, studentName));
    }

    /** Artifact labels go through the same cleaner, so no side-effect label can leak an enum. */
    private List<Artifact> artifacts(List<Artifact> raw) {
        List<Artifact> out = new ArrayList<>();
        for (Artifact a : raw) {
            out.add(new Artifact(a.kind(), DisplayLabels.proof(a.label()), a.value(), a.href()));
        }
        return out;
    }

    /** Raw history rows in, English out. Nothing else formats the trail. */
    private List<TrailEntry> trail(List<RequestHistory> rows, String studentName) {
        List<TrailEntry> out = new ArrayList<>();
        for (RequestHistory h : rows) {
            out.add(new TrailEntry(
                    DisplayLabels.transition(h.fromState, h.toState),
                    DisplayLabels.actor(h.actor, studentName),
                    h.note,
                    DisplayLabels.effects(h.effects),
                    DisplayLabels.proof(h.effectLog),
                    FMT.format(h.at)));
        }
        return out;
    }

    private String tone(WorkflowSpec spec, RequestState state) {
        if (spec.terminalTone().containsKey(state)) return spec.terminalTone().get(state);
        if (spec.offPath().containsKey(state)) return spec.offPath().get(state).tone();
        return "pending";
    }

    /** Type-agnostic. `skipped` is how an automated bypass becomes visible to the student. */
    private List<TimelineStep> steps(WorkflowSpec spec, Request r, List<RequestHistory> rows) {
        Set<RequestState> visited = new HashSet<>();
        for (RequestHistory h : rows) visited.add(h.toState);

        List<TimelineStep> out = new ArrayList<>();
        WorkflowSpec.OffPath off = spec.offPath().get(r.state);

        if (off != null) {
            for (WorkflowSpec.Step s : spec.steps()) {
                out.add(new TimelineStep(s.label(), visited.contains(s.key()) ? "done" : "pending"));
            }
            out.add(new TimelineStep(off.label(), "failed"));
            return out;
        }

        int idx = -1;
        for (int i = 0; i < spec.steps().size(); i++) {
            if (spec.steps().get(i).key() == r.state) { idx = i; break; }
        }
        boolean terminal = spec.isTerminal(r.state);
        for (int i = 0; i < spec.steps().size(); i++) {
            WorkflowSpec.Step s = spec.steps().get(i);
            String status;
            if (i < idx) status = visited.contains(s.key()) ? "done" : "skipped";
            else if (i == idx) status = terminal ? "done" : "current";
            else status = "pending";
            out.add(new TimelineStep(s.label(), status));
        }
        return out;
    }

    private ActionButton studentAction(WorkflowSpec spec, Request r) {
        for (Transition t : spec.from(r.state)) {
            if (t.actor() == Actor.STUDENT) {
                return new ActionButton(t.label(), t.event().name(),
                        t.tone(), t.requiresNote(), t.inputLabel());
            }
        }
        return null;
    }

    /**
     * The COLLAPSED card's one line. Deliberately does NOT carry the engine's effect-log
     * prose — the artifact bullets (serial, verification ID, download, credits, attendance)
     * already show the outcome. Off-path states keep the human-written reason, because that
     * is the one thing a student needs and it is their reviewer's words, not engine output.
     */
    private String headline(WorkflowSpec spec, Request r, RequestPayload payload,
                            List<RequestHistory> rows) {
        RequestHistory last = rows.isEmpty() ? null : rows.get(rows.size() - 1);
        String note = last == null || last.note == null ? "" : last.note.trim();

        if (spec.offPath().containsKey(r.state)) {
            String label = spec.offPath().get(r.state).label();
            return note.isEmpty() ? label + "." : label + " — \u201c" + note + "\u201d";
        }
        String status = DisplayLabels.status(r.state);
        if (status != null) return status;

        String handler = payload.handledBy();
        if (handler == null) {
            handler = spec.from(r.state).stream()
                    .filter(t -> t.actor() != Actor.SYSTEM)
                    .map(t -> t.actor().display())
                    .findFirst().orElse("CampusOS");
        }
        return "Currently with " + handler + ".";
    }

}

File: modules/student-experience-js/src/service/PresentationService.js
import { Actor } from '../domain/enums.js';
import { TransitionMatrix } from '../engine/TransitionMatrix.js';
import { PayloadCodec } from '../payload/PayloadCodec.js';
import { requestHistoryRepo } from '../repo/requestHistoryRepo.js';
import { studentRepo } from '../repo/studentRepo.js';
import { DisplayLabels } from '../view/DisplayLabels.js';

function fmt(iso) {
  const d = new Date(iso);
  return d.toLocaleString('en-GB', { day: 'numeric', month: 'short', hour: '2-digit', minute: '2-digit', hour12: false });
}

function studentName(scope) {
  const s = studentRepo.findByIdAndTenantId(scope.studentId, scope.tenantId);
  return s ? s.name : 'You';
}

function tone(spec, state) {
  if (spec.terminalTone[state]) return spec.terminalTone[state];
  if (spec.offPath[state]) return spec.offPath[state].tone;
  return 'pending';
}

/** Type-agnostic. `skipped` is how an automated bypass becomes visible to the student. */
function steps(spec, r, rows) {
  const visited = new Set(rows.map((h) => h.toState));

  const off = spec.offPath[r.state];
  if (off) {
    const out = spec.steps.map((s) => ({ label: s.label, status: visited.has(s.key) ? 'done' : 'pending' }));
    out.push({ label: off.label, status: 'failed' });
    return out;
  }

  const idx = spec.steps.findIndex((s) => s.key === r.state);
  const terminal = spec.isTerminal(r.state);
  return spec.steps.map((s, i) => {
    let status;
    if (i < idx) status = visited.has(s.key) ? 'done' : 'skipped';
    else if (i === idx) status = terminal ? 'done' : 'current';
    else status = 'pending';
    return { label: s.label, status };
  });
}

/**
 * The COLLAPSED card's one line. Deliberately does NOT carry the engine's effect-log prose —
 * the artifact bullets already show the outcome. Off-path states keep the human-written reason.
 */
function headline(spec, r, payload, rows) {
  const last = rows.length ? rows[rows.length - 1] : null;
  const note = last && last.note ? last.note.trim() : '';

  if (spec.offPath[r.state]) {
    const label = spec.offPath[r.state].label;
    return note === '' ? `${label}.` : `${label} — “${note}”`;
  }
  const status = DisplayLabels.status(r.state);
  if (status) return status;

  let handler = payload.handledBy();
  if (!handler) {
    const edge = spec.from(r.state).find((t) => t.actor !== 'SYSTEM');
    handler = edge ? Actor.display(edge.actor) : 'CampusOS';
  }
  return `Currently with ${handler}.`;
}

function studentAction(spec, r) {
  const edge = spec.from(r.state).find((t) => t.actor === 'STUDENT');
  if (!edge) return null;
  return { label: edge.label, event: edge.event, tone: edge.tone, requiresNote: edge.requiresNote, inputLabel: edge.inputLabel };
}

function artifacts(raw) {
  return raw.map((a) => ({ ...a, label: DisplayLabels.proof(a.label) }));
}

function trail(rows, name) {
  return rows.map((h) => ({
    transition: DisplayLabels.transition(h.fromState, h.toState),
    actor: DisplayLabels.actor(h.actor, name),
    note: h.note,
    effects: DisplayLabels.effects(h.effects),
    proof: DisplayLabels.proof(h.effectLog),
    at: fmt(h.at),
    hasNote: !!(h.note && h.note.trim()),
    hasEffects: !!(h.effects && h.effects.trim()),
    hasProof: !!(h.effectLog && h.effectLog.trim()),
  }));
}

function card(scope, r, name) {
  const spec = TransitionMatrix.spec(r.type);
  const payload = PayloadCodec.read(r.type, r.payload);
  const rows = requestHistoryRepo.findByRequestIdAndTenantIdAndStudentIdOrderByIdAsc(r.id, scope.tenantId, scope.studentId);

  return {
    id: r.id,
    type: r.type,
    typeLabel: spec.label,
    title: payload.title(),
    subtitle: payload.subtitle(),
    state: r.state,
    stateLabel: DisplayLabels.stateLabel(spec.labelFor(r.state), payload.handledBy()),
    badgeTone: tone(spec, r.state),
    headline: headline(spec, r, payload, rows),
    steps: steps(spec, r, rows),
    studentAction: studentAction(spec, r),
    artifacts: artifacts(payload.artifacts()),
    createdAt: fmt(r.createdAt),
    updatedAt: fmt(r.updatedAt),
    trail: trail(rows, name),
    isOpen() { return this.badgeTone !== 'success' && this.badgeTone !== 'danger'; },
  };
}

/**
 * Turns a Request of ANY type into one RequestCard, using only WorkflowSpec metadata and the
 * payload's own polymorphic title()/subtitle()/artifacts(). No switch on RequestType.
 */
export const PresentationService = {
  cards(scope, requests) {
    const name = studentName(scope);
    return requests.map((r) => card(scope, r, name));
  },

  card(scope, r) { return card(scope, r, studentName(scope)); },

  /**
   * THE STAFF RENDERING of a student's cards: same read model, minus links the staff member
   * is not authorized to follow. /documents/{id}/download is STUDENT-only.
   */
  staffCards(scope, requests) {
    return this.cards(scope, requests).map((c) => ({
      ...c,
      artifacts: c.artifacts.map((a) => (a.kind === 'DOCUMENT' ? { ...a, href: null } : a)),
    }));
  },
};
